[v2] tools/cgclassify: add -r option for systemd scope

kamalesh-babulal opened the pull request:
This patch set provides an option for users to replace the default idle_thread
created for systemd scope. It introduces '-r' option, when specified, the first
pid passed as the argument to cgclassify will replace the default idle_thread,
if it exists. Also, if it's the only task, killing it will remove the .scope cgroup it's
running in. It also adds support for ftests to pass the -r flag and a test case to
exercise the different likely and unlikely paths with replace flag.

kamalesh-babulal commented:
**V2 Changes**:
- Rebased the code based on the latest main.
- Addressed the reviewed comments from @drakenclimber #355 
  - Fixed memory leaks from `find_mnt_point()` 
  - Added `else` false positive cases

drakenclimber commented on src/tools/cgclassify.c:
```diff
@@ -198,7 +231,459 @@ int main(int argc, char *argv[])
+	 */
+	if (is_cgroup_mode_unified()) {
+		pthread_rwlock_rdlock(&cg_mount_table_lock);
+		cg_build_path_locked(info[idx].cgrp_path, cgroup_path, NULL);
```
What if a user does something like this:
```
cgcreate -c -S -g cpu:my.slice/db.scope   # set my.slice/db.scope as the default
cgclassify -r -g cpu:my.slice/db.scope 1234
```

This line would have `cg_build_path_locked()` prepend `my.slice/db.scope` in front of `info[idx].cgroup_path`.  But since `info[idx].cgroup_path` came directly from `/proc/{pid}/cgroup`, I don't see why we would ever want to prepend the slice/scope info.  What am I missing?

kamalesh-babulal replied:
In this above example the user is expected to use one of the following commands:
`cgclassify -b -r -g cpu:my.slice/db.scope 1234` or
`cgclassify -r -g cpu:/ 1234`

either the user could use `-b` option to ignore the default delegated `slice/scope` or pass `/` referring to the root a.k.a delegated `slice and scope`

drakenclimber replied:
But since the cgroup path info is coming from `/proc/{pid}/cgroup`, the slice/scope are already there right?  So, wouldn't your first example end up with a path like `/sys/fs/cgroup/my.slice/db.scope/my.slice/db.scope`?

drakenclimber replied:
Gotcha.  I missed the `-b` in your first example. Thanks!

drakenclimber commented on src/tools/cgclassify.c:
```diff
@@ -198,7 +231,459 @@ int main(int argc, char *argv[])
+	for (idx = 0; info[idx].ctrl_name[0] != '\0'; idx++) {
+		/* find the systemd cgroup path */
+		if (!strcmp(info[idx].ctrl_name, "name=systemd")) {
+			find_mnt_point("name=systemd ", &cgrp_proc_path);
```
You have a space at the end of `name=systemd `.  If a system ever places the `name=systemd` option before `noexec`, `nodev`, etc. then it won't be found.  Is this by design?

kamalesh-babulal replied:
@drakenclimber Thank you so much for the review. Agreed, it's not robust will come up with something better.

kamalesh-babulal commented:
**v3 Changes**
- Addressed the review comments from @drakenclimber 
- Fixed a memory leak and uninitialized var in `find_scope_pid()`
- Improved error path return.
